### DISCUSSION TITLE
400: JSON is malformed: invalid character when trying to send binary data

### DISCUSSION THREAD

**Discussion by ghost** [litestar-org/litestar#2762] [Q&A] [answered]
### Description

Hello,

I have a client that sends a pytorch model (after converting it to base64 string), for which I want my litestar server to read it and use it afterwards. 

My client code:

```
import io
import torch
import base64
import requests

model = torch.hub.load('pytorch/vision:v0.10.0', 'alexnet', pretrained=True)
buffer = io.BytesIO()
torch.save(model, buffer)
buffer.seek(0)
model = buffer.read()
model = base64.b64encode(model).decode()
data = {"module_name": "module", "version_name": "version", "model": model}
response = requests.post("http://localhost:8080/api/v1/register", data=data)
print(response)
```

My server code:

```
from typing import Annotated
from dataclasses import dataclass
import io
import torch
import mlflow
import mlflow.pytorch
from litestar import Controller, post


@dataclass
class RegisterRequest:
    module_name: str
    version_name: str
    model: str

class MLFlowController(Controller):
    @post("/register")
    async def register(self, data: RegisterRequest) -> None:
        model = data.model
        # model = base64.b64decode(model)
        # model = torch.load(io.BytesIO(model))
        # mlflow.pytorch.log_model(model, f"{data.module_name}_{data.version_name}")
```

However, my litestar server is throwing the following error:

`litestar.exceptions.http_exceptions.ClientException: 400: JSON is malformed: invalid character (byte 0)`

I tried using dict[str, str] instead of creating a dataclass but it did not help. What should I do for my scenario? Note that sending a pytorch model is just an example; I might send a tensor or anything else that is not a simple class. Also I do not want to use multi-part requests for infra reasons).

### Litestar Version

2 (latest)

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2761">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2761/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2761/pledge.svg">
</picture>
</a>

**Comment by provinzkraut** [comment, marked as answer]
> Also I do not want to use multi-part requests for infra reasons).

But you are. Passing a dict to `data` will cause it to be form encoded. 

From the requests docs:

> `data` – the body to attach to the request. If a dictionary or list of tuples `[(key, value)] is provided`, form-encoding will take place.

https://requests.readthedocs.io/en/latest/api/#requests.Request

The error you are seeing is because you are telling Litestar to decode form encoded data as JSON.

You need to either send your data as JSON (e.g. by encoding it as a string before passing it to the `data` parameter, or passing your dictionary to the `json` parameter instead), or by setting the appropriate content-type in your handler so Litestar can decode it correctly: https://docs.litestar.dev/latest/usage/requests.html#content-type
